### PR TITLE
Implementation of virial stress in sparse kernel model for the GAP kernel

### PR DESCRIPTION
**Pull request by agoscinski**
Implements, tests, and documents kernel stress computation. 

Each part was coded by the person in the brackets and is/was revised by the other person

Implements kernel stress computation (Felix, Alex)
-	src/rascal/representations/calculator_spherical_expansion.hh (Felix)
-	src/rascal/representations/calculator_spherical_invariants.hh (Felix)
-	src/rascal/models/sparse_kernels.hh (Felix)
- src/rascal/structure_managers/atomic_structure.hh (Alex)

Check for mismatching species if-case was AFTER access with species which resulted in access error, now it is before (Alex)
-	src/rascal/models/sparse_points.hh

Implements numerical kernel stress by displacement of strain tensor (Alex)
-	src/rascal/models/numerical_kernel_gradients.hh
-	src/rascal/structure_managers/atomic_structure.hh

Implements tests for kernel stress (Felix)
-	tests/test_sparse_kernels.cc
-	reference_data/tests_only/sparse_kernel_inputs.json

Implements bindings for kernel stress (Felix)
- bindings/bind_py_models.cc
-	bindings/bind_py_models.hh
-	bindings/rascal/models/kernels.py
-	bindings/rascal/models/krr.py
-	bindings/rascal/models/sparse_points.py
-	bindings/rascal/representations/spherical_invariants.py

Implements tests for numerical kernel stress comparing it with the numerical stress of the ase library (Alex). Numerical kernel stress computation on cpp site is verfied with one on the python site which is verified with the numercial stress computation of the `ase` library
-	tests/python/python_binding_tests.py
-	tests/python/python_models_test.py
-	tests/python/test_utils.py

Adds documentation for sparse kernel gradient and stress and reworked variable names to be more verbose and explicit (Alex)
-	src/rascal/models/sparse_kernels.hh
-	docs/source/reference/cpp.rst

Tasks before review:

- [x] Add tests, examples, and complete documentation of any added functionality
    - [x] Make sure the autogenerated documentation appears in Sphinx and is
          formatted correctly (ask @max-veit if you need help with this task).
    - [x] Write additional documentation in the Sphinx (not docstrings!) to
          explain the feature and its usage in plain English
    - [x] Make sure the examples run (!) and produce the expected output
    - [x] For bugfix pull requests, list here which tests have been added or re-enabled
          to verify the fix and catch future regressions as well as similar bugs
          elsewhere in the code.
- [x] Run `make lint` on the project, ensure it passes
    - [x] Run `make pretty-cpp` and `make pretty-python`, check that the
          auto-formatting is sensible
    - [x] Review variable names, make sure they're descriptive (ask a friend)
    - [x] Review all TODOs, convert to issues wherever possible
    - [x] Make sure draft, in-progress, and commented-out code is moved to its
          own branch
- [x] If committing any code in Jupyter/IPython notebooks, install and run `nbstripout`
- [x] Merge with master and resolve any conflicts
- [x] (anything else that's still in progress)

### CODE REVIEW

**Comment by felixmusil** [comment]
I used the stress to run a cell optimization and turns out that the sign should be flipped: the model is trained on energies so it should output `-stress` but it outputs `stress`.

File: requirements.txt [hunk 1]
```diff
@@ -14,7 +14,7 @@ nbsphinx
 pygments>=2.4.1
 # Developers tools - code formatting
 cpplint
-black
+black==20.8b1
```
Comment (Luthaf) [inline requirements.txt:17]: While locking black version is a very good idea, I would rather not lock ourself to a beta version =)

Comment (felixmusil) [reply requirements.txt:17]: black is still in a beta mode so all release have a b in their tag. Or I don't set the version and we get always the latest ?

Comment (Luthaf) [reply requirements.txt:17]: My bad, this works then

**Review by felixmusil — APPROVED** [review]
sounds good to me

**Pull request merged**